### PR TITLE
AF-2856: Create Quarkus Backend for DashBuilder Runtime

### PR DESCRIPTION
**Pull request by jesuino**
**JIRA**: 
https://issues.redhat.com/browse/AF-2852
https://issues.redhat.com/browse/AF-2856

**referenced Pull Requests**: https://github.com/kiegroup/kie-soup/pull/204

This PR creates a new distribution for Dashbuilder Runtime based on Quarkus. The result is a JAR that can be executed using `java -jar` command

Since CI Servers are not yet configured you can download the JAR from here: https://drive.google.com/file/d/1qYKbbbW-KjOpPvb51tw4Yv16h6Tsg-Wo/view?usp=sharing

To run you just need Java 11 and then:
```
java -jar ./dashbuilder-runtime-app-7.56.0-SNAPSHOT-runner.jar
```

The only big differences to the old DB Runtime are:

* There's no more authentication;
* SQL Datasources setup are now done using system properties.  You can run a SQL dashboard using this ZIP: https://github.com/jesuino/dashbuilder-data/blob/master/dashboards/Orders.zip?raw=true

The SQL with data can be downloaded from this link: https://www.w3resource.com/sql/sql-table.php

So the steps to run a SQL Dashboard:

1) Create a database in your system of preference 
2) Load the SQL above in your database;
3) Run DashBuilder Runtime with the system properties to setup the SQL database.

```
java -Ddashbuilder.runtime.multi=true \
-Ddashbuilder.datasources=sample \
-Ddashbuilder.datasource.sample.jdbcUrl=jdbc:mariadb://localhost:3306/sample \
-Ddashbuilder.datasource.sample.providerClassName=org.mariadb.jdbc.Driver \
-Ddashbuilder.datasource.sample.maxSize=10 \
-Ddashbuilder.datasource.sample.principal=repasse \
-Ddashbuilder.datasource.sample.credential=repasse \
-jar target/dashbuilder-runtime-app-7.56.0-SNAPSHOT-runner.jar
```

Make sure to change JDBC URL and Driver according to the used DB. The supported databases are the same as supported by Agroal from Quarkus 1.11.6:

* H2
* MS SQL
* MariaDB
* MySQL
* Derby
* Postgres

Other sample dashboards can be found here: https://github.com/jesuino/dashbuilder-data/tree/master/dashboards

Everything should simply work with DB as it used to be, if you see any issue please let me know.

Notice that it is not a complete rewrite, so I moved most classes from older DB Runtime to this new dist, that's the reason many copyright headers are using year "2020", because I moved the classes from old DB.  In this change we also separated the client from the backend, now the client can have a plugabble backend. In any case there's no Quarkus Errai adapter, so we build the client separately and then run with DB Runtime Quarkus App. 

Finally you will see an exception threw in the logs. It was corrected recently and it happens due HTTP compression.

```
2021-06-15 12:41:09,005 WARN  [io.net.cha.ChannelOutboundBuffer] (vert.x-eventloop-thread-12) Failed to mark a promise as success because it has failed already: DefaultChannelPromise@420d68bf(failure: io.netty.handler.codec.EncoderException: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1), unnotified cause:: io.netty.handler.codec.EncoderException: io.netty.util.IllegalReferenceCountException: refCnt: 0, decrement: 1
```

If you run Dashbuilder without HTTP compression you will not see the exception anymore:
```
java -Dquarkus.http.enable-compression=false -jar target/dashbuilder-runtime-app-7.56.0-SNAPSHOT-runner.jar
```

### CODE REVIEW

**Review by caponetto — COMMENTED** [review]
Hi @jesuino, thanks for the PR!
I've left some comments, please check if they make sense.

Also, it seems that CI is not configured for this `v8` branch.
Do you know anything about it?
Asking because it would be good to see the output of the build with these changes.

**Comment by sthundat** [comment]
Jenkins run fdb

**Comment by sthundat** [comment]
Hi @jesuino ,

Thanks for this PR!

I have checked the dashbuilder runtime with Quarkus backend using mariadb and everything working as expected.I am able to bring up Orders dashboard in dashbuilder runtime as shown below.
Great work William!

ordersdashboard
 
NOTE:As discussed, I shall do the extensive testing including all relevant DBs when this get productized.

Thanks,
Savitha

File: dashbuilder/dashbuilder-runtime-parent/pom.xml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  <artifactId>dashbuilder-runtime-parent</artifactId>
+  <version>7.56.0-SNAPSHOT</version>
+  <packaging>pom</packaging>
+
+  <name>Dashbuilder Runtime Project</name>
+
+  <modules>
+    <module>dashbuilder-runtime-shared</module>
+    <module>dashbuilder-runtime-client</module>
+    <module>dashbuilder-runtime-app</module>
```
[the remark this answers is not in the document]
Comment (jesuino) [reply dashbuilder/dashbuilder-runtime-parent/pom.xml:38]: Hello @caponetto 

in this very specific case we are making db runtime out of sync with parent BOM. I will send a new commit setting the correct version, which is 8.0.0-SNAPSHOT. In this case we will keep DB Runtime only on this new branch (v8) and manually release it - and keep doing rebase with master

That's why we won't update the parent POM this time!

File: dashbuilder/dashbuilder-runtime-parent/dashbuilder-runtime-app/README.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+./mvnw compile quarkus:dev
+```
+
+## Packaging and running the application
+
+The application can be packaged using:
+```shell script
+./mvnw package
+```
+It produces the `dashbuilder-runtime-quarkus-1.0.0-SNAPSHOT-runner.jar` file in the `/target` directory.
```
[the remark this answers is not in the document]
Comment (jesuino) [reply dashbuilder/dashbuilder-runtime-parent/dashbuilder-runtime-app/README.md:20]: updated the README - thanks!

File: dashbuilder/dashbuilder-runtime-parent/dashbuilder-runtime-app/src/main/java/org/dashbuilder/backend/resources/marshalling/DataSetWriter.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    @Override
+    public void writeTo(DataSet t,
+                        Class<?> type,
+                        Type genericType,
+                        Annotation[] annotations,
+                        MediaType mediaType,
+                        MultivaluedMap<String, Object> httpHeaders,
+                        OutputStream entityStream) throws IOException, WebApplicationException {
+        var dataSetJSONMarshaller = DataSetJSONMarshaller.get();
+        var json = dataSetJSONMarshaller.toJson(t).toJson();
```
[the remark this answers is not in the document]
Comment (jesuino) [reply dashbuilder/dashbuilder-runtime-parent/dashbuilder-runtime-app/src/main/java/org/dashbuilder/backend/resources/marshalling/DataSetWriter.java:58]: found that `toString` works as well!

**Comment by jesuino** [comment]
Hello @sthundat 

Thanks for your feedback - glad to know everything worked!


@caponetto 

Thanks again for another great review.

I just finished the changes. Please let me know if I am missing something.

Thanks!

File: dashbuilder/dashbuilder-runtime-parent/pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (caponetto) [reply dashbuilder/dashbuilder-runtime-parent/pom.xml:38]: Thanks for the info! :)

**Review by caponetto — APPROVED** [review]
Thanks for taking a look at my comments, @jesuino! Great work!

**Review by sthundat — APPROVED** [review]

**Review by ederign — APPROVED** [review]

**Pull request merged**